sklearn/utils/_isfinite.pyx

```diff
@@ -0,0 +1,35 @@
+# Author: jakirkham, Meekail Zain, Thomas Fan
+
+from libc.math cimport isnan, isinf
+cimport cython
+
+cpdef enum FiniteStatus:
+    all_finite = 0
+    has_nan = 1
+    has_infinite = 2
+
+def cy_isfinite(cython.floating[::1] a, bint allow_nan=False):
+    cdef int result
+    with nogil:
+        result = _isfinite(a, allow_nan)
+    return result
+
+cdef inline int _isfinite(cython.floating[::1] a, bint allow_nan) nogil:
+    if allow_nan:
+        return _isfinite_allow_nan(a)
+    else:
+        return _isfinite_disable_nan(a)
+
+cdef inline int _isfinite_allow_nan(cython.floating[::1] a) nogil:
+    for i in range(len(a)):
+        if isinf(a[i]) != 0:
+            return FiniteStatus.has_infinite
+    return FiniteStatus.all_finite
+
+cdef inline int _isfinite_disable_nan(cython.floating[::1] a) nogil:
+    for i in range(len(a)):
```

Comment:
Doesn't it help to use an openmp parallel loop with `prange` here?

Reply:
Interestingly enough, it looks like it performs _worse_ with `prange` -- at least when `num_threads` is set to `None` and auto-resolved by OpenMP. Will test w/ single-threaded prange, but that had [problems](https://github.com/scikit-learn/scikit-learn/pull/23197#issuecomment-1113565644) previously.

![b609ce60-154c-4090-b2ed-eeffb77bbc3b](https://user-images.githubusercontent.com/34613774/174631895-d3139152-64fc-4f6e-b750-0267726ad2e9.png)